proposals/module-types/Example-SharedEverythingDynamicLinking.md

```diff
@@ -0,0 +1,525 @@
+with `malloc`, `free`, etc. Thus, `libc.wasm` is roughly shaped:
+
+```wasm
+;; libc.wasm
+(module
+  (memory (export "memory") 1)
+  (func (export "malloc") (param i32) (result i32) ...impl)
+  ...
+)
+```
+
+The SDK also contains standard library headers which contain declarations
+compatible with `libc.wasm`. To implement them, we first need some helper
+macros, which we'll conveniently define in `stddef.h`:
+
+```c
+/* stddef.h */
+#define WASM_IMPORT(module,name) __attribute__((import_module(#module), import_name(#name))) name
+#define WASM_EXPORT(name) __attribute__((export_name(#name))) name
+```
+
+These macros use the clang-specific attributes [`import_module`], [`import_name`]
+and [`export_name`] to ensure that the annotated C functions produce the
+correct wasm import or export definitions. Other compilers would use their own
+magic syntax to achieve the same effect. Using these macros, we can declare
+`malloc` in `stdlib.h`:
+
+```c
+/* stdlib.h */
+#include <stddef.h>
+#define LIBC(name) WASM_IMPORT(libc, name)
+void* LIBC(malloc)(size_t n);
+```
+
+## libzip
+
+With `libc` sketched out, we have a functional SDK we can now use to create the
+shared library `libzip`. Its C definition has the rough shape:
+
+```c
+/* libzip.h */
+#include <stddef.h>
+#define LIBZIP(name) WASM_IMPORT(libzip, name)
+void* LIBZIP(zip)(void* in, size_t in_size, size_t* out_size);
+```
+```c
+/* libzip.c */
+#include <stdlib.h>
+void* WASM_EXPORT(zip)(void* in, size_t in_size, size_t* out_size) {
+  ...
+  void *p = malloc(n);
+  ...
+}
+```
+
+Note that `libzip.h` annotates the `zip` declaration with an *import* attribute
+so that client modules generate proper wasm *import definitions* while `libzip.c`
+annotates the `zip` definition with an *export* attribute so this function generates
+a proper wasm *export definition*. Compiling with `clang -shared libzip.c`
+produces a `libzip.wasm` of the form:
+
+```wasm
+;; libzip.wat
+(module
+  (import "libc" (instance $libc
+    (export "memory" (memory 1))
+    (export "malloc" (func $malloc (param i32) (result i32)))
+  ))
+  (func (export "zip") (param i32 i32 i32) (result i32)
+    ...
+    call $libc.$malloc
+    ...
+  )
+)
+```
+
+## zipper
+
+Using `libc` and `libzip`, we can now write the `zipper` program:
+
+```c
+/* zipper.c */
+#include <stdlib.h>
+#include "libzip.h"
+int main(int argc, char* argv[]) {
+  ...
+  void *in = malloc(n);
+  ...
+  void *out = zip(in, n, &out_size);
+  ...
+}
+```
+
+When compiled with `clang zipper.c`, we get a `zipper.wasm` shaped like:
+
+```wasm
+;; zipper.wat
+(module
+  (type $Libc (instance
+    (export "memory" (memory 1))
+    (export "malloc" (func $malloc (param i32) (result i32)))
+  ))
+  (type $LibZip (instance
+    (export "zip" (func $zip (param i32 i32 i32) (result i32)))
+  ))
+
+  (import "libc" (module $LIBC
+    (exports $Libc)
+  ))
+  (import "libzip" (module $LIBZIP
+    (import "libc" (instance (type $Libc)))
+    (exports $LibZip)
+  ))
```

Comment:
One perhaps interesting point of note here is that this `import` directive is not fully known at compile time, it's only known fully at link-time. The `libzip.h` header file doesn't mention that imports `libc.wasm`, but something needs to determine that `libzip.wasm` does indeed import `libc.wasm`. This is a bit of a silly case with `libc.wasm`, but the problem in general of knowning transitively what's required to instantiate a library is only known from looking at the final `*.wasm` file.

This I think means that the final linker, LLD typically, would have to be pretty smart about loading `libzip.wasm` and reading its transitive closure of dependencies to list here too.

This sort of leads into my next comment though...

Reply:
I think it might be a bit simpler than that: each individual object file will contain some subset of imports from `libc`.  (Either as instance imports or as imports of individual fields.)  LLD just collects these together and can produce an instance import by just merging multiple field imports (merging `(import "libc" "x" ...)` and `(import "libc" "y" ...)` into `(import "libc" (instance (export "x") (export "y")))`.